### PR TITLE
ORCA: Pruning the output columns in CTE

### PR DESCRIPTION
**Pull request by jiaqizho**
Fixes #ISSUE_Number

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Test Plan

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact

**Performance:**

**User-facing changes:**

**Dependencies:**

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### CI Skip Instructions

---

### CODE REVIEW

**Comment by jiaqizho** [comment]
i will add the tests later...

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 1]
```diff
@@ -4153,6 +4163,17 @@ CTranslatorDXLToPlStmt::TranslateDXLCTEConsumerToSharedScan(
...
 	for (ULONG ul = 0; ul < num_of_proj_list_elem; ul++)
 	{
+		AttrNumber varattno = (AttrNumber)ul + 1;
+		if (output_colidx_map) {
+			ULONG remapping_idx;
+			remapping_idx = *(*output_colidx_map)[ul];
+			if (producer_colidx_map) {
+				remapping_idx = *(*producer_colidx_map)[remapping_idx];
+			}
+			GPOS_ASSERT(remapping_idx != ULONG_MAX);
```
Comment (jiaqizho) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: gpos::ulong_max

**Comment by my-ship-it** [comment]
We need to be careful about VOLATILE and STABLE functions, for example:

```
create table t1(v1 int, v2 int, v3 int);
insert into t1 values(generate_series(1, 10), generate_series(11, 20), generate_series(21, 30));
analyze t1;

create table t2(v1 int, v2 int, v3 int);
insert into t2 values(generate_series(0, 100), generate_series(100, 200), generate_series(200, 300));

-- should pruned both seq scan and shared scan
postgres=# explain verbose with c1 as (select v1, v2, random() v3 from t1) select c11.v1 from c1 as c11 left join c1 as c22 on c11.v1=c22.v1 where c11.v1 < 5;
                                             QUERY PLAN
----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1293.00 rows=5 width=4)
   Output: share0_ref3.v1
   ->  Sequence  (cost=0.00..1293.00 rows=2 width=4)
         Output: share0_ref3.v1
         ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=1)
               Output: share0_ref1.v1
               ->  Seq Scan on public.t1  (cost=0.00..431.00 rows=2 width=4)
                     Output: t1.v1
                     Filter: (t1.v1 < 5)
         ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=4)
               Output: share0_ref3.v1
               Hash Cond: (share0_ref3.v1 = share0_ref2.v1)
               ->  Result  (cost=0.00..431.00 rows=2 width=4)
                     Output: share0_ref3.v1
                     Filter: (share0_ref3.v1 < 5)
                     ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
                           Output: share0_ref3.v1
               ->  Hash  (cost=431.00..431.00 rows=2 width=4)
                     Output: share0_ref2.v1
                     ->  Result  (cost=0.00..431.00 rows=2 width=4)
                           Output: share0_ref2.v1
                           Filter: (share0_ref2.v1 < 5)
                           ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
                                 Output: share0_ref2.v1
 Optimizer: GPORCA
(25 rows)
```

It is obvious we shouldn't prune v3 out safely here.

**Comment by jiaqizho** [comment]
> We need to be careful about VOLATILE and STABLE functions, for example:
> 
> ```
> create table t1(v1 int, v2 int, v3 int);
> insert into t1 values(generate_series(1, 10), generate_series(11, 20), generate_series(21, 30));
> analyze t1;
> 
> create table t2(v1 int, v2 int, v3 int);
> insert into t2 values(generate_series(0, 100), generate_series(100, 200), generate_series(200, 300));
> 
> -- should pruned both seq scan and shared scan
> postgres=# explain verbose with c1 as (select v1, v2, random() v3 from t1) select c11.v1 from c1 as c11 left join c1 as c22 on c11.v1=c22.v1 where c11.v1 < 5;
>                                              QUERY PLAN
> ----------------------------------------------------------------------------------------------------
>  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1293.00 rows=5 width=4)
>    Output: share0_ref3.v1
>    ->  Sequence  (cost=0.00..1293.00 rows=2 width=4)
>          Output: share0_ref3.v1
>          ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=1)
>                Output: share0_ref1.v1
>                ->  Seq Scan on public.t1  (cost=0.00..431.00 rows=2 width=4)
>                      Output: t1.v1
>                      Filter: (t1.v1 < 5)
>          ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=4)
>                Output: share0_ref3.v1
>                Hash Cond: (share0_ref3.v1 = share0_ref2.v1)
>                ->  Result  (cost=0.00..431.00 rows=2 width=4)
>                      Output: share0_ref3.v1
>                      Filter: (share0_ref3.v1 < 5)
>                      ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
>                            Output: share0_ref3.v1
>                ->  Hash  (cost=431.00..431.00 rows=2 width=4)
>                      Output: share0_ref2.v1
>                      ->  Result  (cost=0.00..431.00 rows=2 width=4)
>                            Output: share0_ref2.v1
>                            Filter: (share0_ref2.v1 < 5)
>                            ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
>                                  Output: share0_ref2.v1
>  Optimizer: GPORCA
> (25 rows)
> ```
> 
> It is obvious we shouldn't prune v3 out safely here.

In fact, this is what ORCA's prune logic (specifically the `used/unused` status setting for CColRef) does. 

running follow SQL in **MAIN branch** :

ex.
```

explain verbose select c11.v1 from (select v1, v2, random() v3 from t1) as c11 left join (select v1, v2, v3 from t1) as c22 on c11.v1=c22.v1 where c11.v1 < 5;

                                    QUERY PLAN
----------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=5 width=4)
   Output: t1.v1
   ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=4)
         Output: t1.v1
         Hash Cond: (t1.v1 = t1_1.v1)
         ->  Seq Scan on public.t1  (cost=0.00..431.00 rows=2 width=4)
               Output: t1.v1
               Filter: (t1.v1 < 5)
         ->  Hash  (cost=431.00..431.00 rows=2 width=4)
               Output: t1_1.v1
               ->  Seq Scan on public.t1 t1_1  (cost=0.00..431.00 rows=2 width=4)
                     Output: t1_1.v1
                     Filter: (t1_1.v1 < 5)
 Settings: optimizer = 'on'
 Optimizer: GPORCA
(15 rows)

CREATE OR REPLACE FUNCTION emp_output()
RETURNS TEXT
LANGUAGE plpgsql
AS $$
BEGIN
    RETURN 'emp';
END;
$$;

explain verbose select c11.v1 from (select v1, v2, emp_output() v3 from t1) as c11 left join (select v1, v2, v3 from t1) as c22 on c11.v1=c22.v1 where c11.v1 < 5;
                                    QUERY PLAN
----------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=5 width=4)
   Output: t1.v1
   ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=4)
         Output: t1.v1
         Hash Cond: (t1.v1 = t1_1.v1)
         ->  Seq Scan on public.t1  (cost=0.00..431.00 rows=2 width=4)
               Output: t1.v1
               Filter: (t1.v1 < 5)
         ->  Hash  (cost=431.00..431.00 rows=2 width=4)
               Output: t1_1.v1
               ->  Seq Scan on public.t1 t1_1  (cost=0.00..431.00 rows=2 width=4)
                     Output: t1_1.v1
                     Filter: (t1_1.v1 < 5)
 Settings: optimizer = 'on'
 Optimizer: GPORCA
(15 rows)
```

The function always been pruned.

**Comment by my-ship-it** [comment]
> create table t1(v1 int, v2 int, v3 int);
> insert into t1 values(generate_series(1, 10), generate_series(11, 20), generate_series(21, 30));
> analyze t1;

It seems like a bug of ORCA.  For planner of PG, it produces, where random is volatile
```
postgres=# select proname, provolatile from pg_proc where proname = 'random';
 proname | provolatile
---------+-------------
 random  | v
(1 row)


postgres=# explain verbose select c11.v1 from (select v1, v2, random() v3 from t1) as c11 left join (select v1, v2, v3 from t1) as c22 on c11.v1=c22.v1 where c11.v1 < 5
;
                                      QUERY PLAN
---------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.07..2.17 rows=3 width=4)
   Output: c11.v1
   ->  Hash Right Join  (cost=1.07..2.12 rows=1 width=4)
         Output: c11.v1
         Hash Cond: (t1.v1 = c11.v1)
         ->  Seq Scan on public.t1  (cost=0.00..1.03 rows=3 width=4)
               Output: t1.v1, t1.v2, t1.v3
         ->  Hash  (cost=1.06..1.06 rows=1 width=4)
               Output: c11.v1
               ->  Subquery Scan on c11  (cost=0.00..1.06 rows=1 width=4)
                     Output: c11.v1
                     ->  Seq Scan on public.t1 t1_1  (cost=0.00..1.04 rows=1 width=16)
                           Output: t1_1.v1, NULL::integer, random()
                           Filter: (t1_1.v1 < 5)
 Settings: optimizer = 'off'
 Optimizer: Postgres query optimizer
(16 rows)
```

But for emp_output, it produces:
```
postgres=# select proname, provolatile from pg_proc where proname = 'emp_output';
-[ RECORD 1 ]-----------
proname     | emp_output
provolatile | i

postgres=# explain verbose select c11.v1 from (select v1, v2, emp_output() v3 from t1) as c11 left join (select v1, v2, v3 from t1) as c22 on c11.v1=c22.v1 where c11.v1 < 5
;
                                 QUERY PLAN
----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=1.06..2.16 rows=3 width=4)
   Output: t1.v1
   ->  Hash Right Join  (cost=1.06..2.11 rows=1 width=4)
         Output: t1.v1
         Hash Cond: (t1_1.v1 = t1.v1)
         ->  Seq Scan on public.t1 t1_1  (cost=0.00..1.03 rows=3 width=4)
               Output: t1_1.v1, t1_1.v2, t1_1.v3
         ->  Hash  (cost=1.04..1.04 rows=1 width=4)
               Output: t1.v1
               ->  Seq Scan on public.t1  (cost=0.00..1.04 rows=1 width=4)
                     Output: t1.v1
                     Filter: (t1.v1 < 5)
 Settings: optimizer = 'off'
 Optimizer: Postgres query optimizer
(14 rows)
```

An issue has beend created: https://github.com/apache/cloudberry/issues/1130

**Comment by leborchuk** [comment]
Are there any descriptions for understanding what type of pruning we are going to execute? Maybe there are bug/feature notes that describe what type of issue we are working on.

I see a set of queries and some TPC-DS queries in src/test/regress/sql/cte_prune.sql. Is this a complete set? How can we find out if we have missed any corner cases? Also, is the pruning cost-based or not? If it is cost-based, maybe it would be great to see this behaviour in tests.

**Comment by jiaqizho** [comment]
> Are there any descriptions for understanding what type of pruning we are going to execute? Maybe there are bug/feature notes that describe what type of issue we are working on.
> 
> I see a set of queries and some TPC-DS queries in src/test/regress/sql/cte_prune.sql. Is this a complete set? How can we find out if we have missed any corner cases? Also, is the pruning cost-based or not? If it is cost-based, maybe it would be great to see this behaviour in tests.

@leborchuk the case defined in `cte_prune.sql`, 

" Is this a complete set? How can we find out if we have missed any corner cases?"

The following is the idea of ​​the current PR:
1. During the ORCA DXL to expression process, ORCA will prune `UNUSED` output columns in each expression.
2. Before the PR change, the producer and consumer output columns of CTE will be marked as used.
3. The current PR will recalculate the columns that should be output by CTE producer and consumer. Mark them as used.

The above process happens before preprocessing and logical transformation. Therefore, ORCA can **SAFELY** prune unnecessary output columns. For the current PR, instead of processing specific cases, CTE is processed uniformly.

"Also, is the pruning cost-based or not? If it is cost-based, maybe it would be great to see this behaviour in tests."

Current PR process the prune CTE output column before preprocessing and logical transformation. The cost model is not relevant here.

**Comment by leborchuk** [comment]
I generated and checked various queries with complex references - they all worked great for me.

Here what I checked:
```
CREATE TABLE t2 (v1 int, v2 int, v3 int)
  WITH (appendoptimized=true, orientation=column)
  DISTRIBUTED BY (v1);

insert into t2 values(generate_series(0, 100), generate_series(100, 200), generate_series(200, 300));

analyze t2;

// reference to 2nd column

explain verbose with c1 as (select v1, v2, v3 from t2) select c11.v1 from c1 as c11 left join c1 as c22 on c11.v1=c22.v1 where c11.v1 < (select min(v2)/20 from c1);

-----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1324898.20 rows=101 width=4)
   Output: share0_ref3.v1
   ->  Sequence  (cost=0.00..1324898.20 rows=34 width=4)
         Output: share0_ref3.v1
         ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=34 width=1)
               Output: share0_ref1.v1, share0_ref1.v2
               ->  Seq Scan on cte_prune.t2  (cost=0.00..431.00 rows=34 width=8)
                     Output: t2.v1, t2.v2
         ->  Hash Left Join  (cost=0.00..1324467.20 rows=34 width=4)
               Output: share0_ref3.v1
               Hash Cond: (share0_ref3.v1 = share0_ref2.v1)
               ->  Nested Loop  (cost=0.00..1324036.18 rows=34 width=4)
                     Output: share0_ref3.v1
                     Join Filter: (share0_ref3.v1 < ((min(share0_ref4.v2) / 20)))
                     ->  Broadcast Motion 1:3  (slice2)  (cost=0.00..431.00 rows=1 width=4)
                           Output: ((min(share0_ref4.v2) / 20))
                           ->  Finalize Aggregate  (cost=0.00..431.00 rows=1 width=4)
                                 Output: (min(share0_ref4.v2) / 20)
                                 ->  Gather Motion 3:1  (slice3; segments: 3)  (cost=0.00..431.00 rows=1 width=4)
                                       Output: (PARTIAL min(share0_ref4.v2))
                                       ->  Partial Aggregate  (cost=0.00..431.00 rows=1 width=4)
                                             Output: PARTIAL min(share0_ref4.v2)
                                             ->  Shared Scan (share slice:id 3:0)  (cost=0.00..431.00 rows=34 width=4)
                                                   Output: share0_ref4.v1, share0_ref4.v2
                     ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=34 width=4)
                           Output: share0_ref3.v1, share0_ref3.v2
               ->  Hash  (cost=431.00..431.00 rows=34 width=4)
                     Output: share0_ref2.v1, share0_ref2.v2
                     ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=34 width=4)
                           Output: share0_ref2.v1, share0_ref2.v2

// distinct - cannot remove reference

explain verbose with c1 as (select distinct v1, v2, v3 from t2) select c11.v1 from c1 as c11 left join c1 as c22 on c11.v1=c22.v1 where c11.v1 < 5;

 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1293.00 rows=6 width=4)
   Output: share0_ref3.v1
   ->  Sequence  (cost=0.00..1293.00 rows=2 width=4)
         Output: share0_ref3.v1
         ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=1)
               Output: share0_ref1.v1
               ->  GroupAggregate  (cost=0.00..431.00 rows=2 width=4)
                     Output: t2.v1
                     Group Key: t2.v1, t2.v2, t2.v3
                     ->  Sort  (cost=0.00..431.00 rows=2 width=12)
                           Output: t2.v1, t2.v2, t2.v3
                           Sort Key: t2.v1, t2.v2, t2.v3
                           ->  Seq Scan on cte_prune.t2  (cost=0.00..431.00 rows=2 width=12)
                                 Output: t2.v1, t2.v2, t2.v3
                                 Filter: (t2.v1 < 5)
         ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=4)
               Output: share0_ref3.v1
               Hash Cond: (share0_ref3.v1 = share0_ref2.v1)
               ->  Result  (cost=0.00..431.00 rows=2 width=4)
                     Output: share0_ref3.v1
                     Filter: (share0_ref3.v1 < 5)
                     ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
                           Output: share0_ref3.v1
               ->  Hash  (cost=431.00..431.00 rows=2 width=4)
                     Output: share0_ref2.v1
                     ->  Result  (cost=0.00..431.00 rows=2 width=4)
                           Output: share0_ref2.v1
                           Filter: (share0_ref2.v1 < 5)
                           ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
                                 Output: share0_ref2.v1

// order by column not in select list

explain verbose with c1 as (select v1, v2, v3 from t2) select c11.v1 from c1 as c11 left join c1 as c22 on c11.v1=c22.v1 where c11.v1 < 5 order by c11.v2 desc;

                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Result  (cost=0.00..1293.00 rows=6 width=4)
   Output: share0_ref3.v1
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1293.00 rows=6 width=4)
         Output: share0_ref3.v1, share0_ref3.v2
         Merge Key: share0_ref3.v2
         ->  Sort  (cost=0.00..1293.00 rows=2 width=8)
               Output: share0_ref3.v1, share0_ref3.v2
               Sort Key: share0_ref3.v2 DESC
               ->  Sequence  (cost=0.00..1293.00 rows=2 width=8)
                     Output: share0_ref3.v1, share0_ref3.v2
                     ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=1)
                           Output: share0_ref1.v1, share0_ref1.v2
                           ->  Seq Scan on cte_prune.t2  (cost=0.00..431.00 rows=2 width=8)
                                 Output: t2.v1, t2.v2
                                 Filter: (t2.v1 < 5)
                     ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=8)
                           Output: share0_ref3.v1, share0_ref3.v2
                           Hash Cond: (share0_ref3.v1 = share0_ref2.v1)
                           ->  Result  (cost=0.00..431.00 rows=2 width=8)
                                 Output: share0_ref3.v1, share0_ref3.v2
                                 Filter: (share0_ref3.v1 < 5)
                                 ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=8)
                                       Output: share0_ref3.v1, share0_ref3.v2
                           ->  Hash  (cost=431.00..431.00 rows=2 width=4)
                                 Output: share0_ref2.v1
                                 ->  Result  (cost=0.00..431.00 rows=2 width=4)
                                       Output: share0_ref2.v1
                                       Filter: (share0_ref2.v1 < 5)
                                       ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
                                             Output: share0_ref2.v1, share0_ref2.v2

// chained with-clause

with c1 as (select v1, v2, v3 from t2),
c2 as (select v1 as v2, v2 as v1 from c1), /* switch names */
c3 as (select v1 as v2, v2 as v1 from c2) /* rename back */
 select c11.v1 from c3 as c11 left join c3 as c22 on c11.v1=c22.v1 where c11.v1 < 5 order by c11.v2 desc;


 Result  (cost=0.00..1293.00 rows=6 width=4)
   Output: share2_ref3.v1
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1293.00 rows=6 width=4)
         Output: share2_ref3.v1, share2_ref3.v2
         Merge Key: share2_ref3.v2
         ->  Sort  (cost=0.00..1293.00 rows=2 width=8)
               Output: share2_ref3.v1, share2_ref3.v2
               Sort Key: share2_ref3.v2 DESC
               ->  Sequence  (cost=0.00..1293.00 rows=2 width=8)
                     Output: share2_ref3.v1, share2_ref3.v2
                     ->  Shared Scan (share slice:id 1:2)  (cost=0.00..431.00 rows=2 width=1)
                           Output: share2_ref1.v2, share2_ref1.v1
                           ->  Seq Scan on cte_prune.t2  (cost=0.00..431.00 rows=2 width=8)
                                 Output: t2.v2, t2.v1
                                 Filter: (t2.v1 < 5)
                     ->  Hash Left Join  (cost=0.00..862.00 rows=2 width=8)
                           Output: share2_ref3.v1, share2_ref3.v2
                           Hash Cond: (share2_ref3.v1 = share2_ref2.v1)
                           ->  Result  (cost=0.00..431.00 rows=2 width=8)
                                 Output: share2_ref3.v2, share2_ref3.v1
                                 Filter: (share2_ref3.v1 < 5)
                                 ->  Shared Scan (share slice:id 1:2)  (cost=0.00..431.00 rows=2 width=8)
                                       Output: share2_ref3.v2, share2_ref3.v1
                           ->  Hash  (cost=431.00..431.00 rows=2 width=4)
                                 Output: share2_ref2.v1
                                 ->  Result  (cost=0.00..431.00 rows=2 width=4)
                                       Output: share2_ref2.v1
                                       Filter: (share2_ref2.v1 < 5)
                                       ->  Shared Scan (share slice:id 1:2)  (cost=0.00..431.00 rows=2 width=4)
                                             Output: share2_ref2.v2, share2_ref2.v1

// lateral join

not supported by orca
```

**Review by leborchuk — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Review by tuhaihe — APPROVED** [review]

File: contrib/pax_storage/src/test/regress/expected/gporca_optimizer.out [hunk 2]
```diff
@@ -11280,10 +11280,9 @@ EXPLAIN (COSTS OFF) WITH abc AS (SELECT onetimefilter1.a, onetimefilter1.b FROM
                        Filter: (share0_ref3.b = f1.b)
                        ->  Materialize
                              ->  Broadcast Motion 3:3  (slice6; segments: 3)
-                                   ->  Result
```
Comment (jiaqizho) [inline contrib/pax_storage/src/test/regress/expected/gporca_optimizer.out:11283]: hold for commit message
